downstream.go

```diff
@@ -891,45 +891,75 @@ func (dc *downstreamConn) handleMessageRegistered(msg *irc.Message) error {
 				Command: irc.ERR_USERSDONTMATCH,
 				Params:  []string{dc.nick, "Cannot change mode for other users"},
 			}}
-		} else if modeStr != "" {
-			params := []string{upstreamName, modeStr}
-			params = append(params, msg.Params[2:]...)
-			uc.SendMessage(&irc.Message{
-				Command: "MODE",
-				Params:  params,
-			})
 		} else {
-			ch, ok := uc.channels[upstreamName]
-			if !ok {
-				return ircError{&irc.Message{
-					Command: irc.ERR_NOSUCHCHANNEL,
-					Params:  []string{dc.nick, name, "No such channel"},
-				}}
-			}
+			if modeStr != "" {
+				params := []string{upstreamName, modeStr}
+				params = append(params, msg.Params[2:]...)
+				uc.SendMessage(&irc.Message{
+					Command: "MODE",
+					Params:  params,
+				})
+			} else {
+				ch, ok := uc.channels[upstreamName]
+				if !ok {
+					return ircError{&irc.Message{
+						Command: irc.ERR_NOSUCHCHANNEL,
+						Params:  []string{dc.nick, name, "No such channel"},
+					}}
+				}
 
-			if ch.modes == nil {
-				// we haven't received the initial RPL_CHANNELMODEIS yet
-				// ignore the request, we will broadcast the modes later when we receive RPL_CHANNELMODEIS
-				return nil
-			}
+				if ch.modes == nil {
+					// we haven't received the initial RPL_CHANNELMODEIS yet
+					// ignore the request, we will broadcast the modes later when we receive RPL_CHANNELMODEIS
+					return nil
+				}
 
-			modeStr, modeParams := ch.modes.Format()
-			params := []string{dc.nick, name, modeStr}
-			params = append(params, modeParams...)
+				modeStr, modeParams := ch.modes.Format()
+				params := []string{dc.nick, name, modeStr}
+				params = append(params, modeParams...)
 
+				dc.SendMessage(&irc.Message{
+					Prefix:  dc.srv.prefix(),
+					Command: irc.RPL_CHANNELMODEIS,
+					Params:  params,
+				})
+			}
+		}
+	case "NAMES":
+		if len(msg.Params) == 0 {
 			dc.SendMessage(&irc.Message{
 				Prefix:  dc.srv.prefix(),
-				Command: irc.RPL_CHANNELMODEIS,
-				Params:  params,
+				Command: irc.RPL_ENDOFNAMES,
+				Params:  []string{dc.nick, "*", "End of /NAMES list"},
 			})
+			return nil
+		}
+
+		channels := strings.Split(msg.Params[0], ",")
+		for _, channel := range channels {
+			uc, upstreamChannel, err := dc.unmarshalEntity(channel)
+			if err != nil {
+				return err
+			}
+
+			ch, err := uc.getChannel(upstreamChannel)
```

Comment:
I prefer to use `ch, ok := uc.channels[upstreamChannel]` when we want to handle both cases (the case where the channel exists, and the case where it doesn't). `getChannel` is just a helper to unify "channel not found" errors.

Reply:
(TBH we could replace `getChannel` with an `errNoSuchChannel` variable)